lib/auth/webauthn/messages_test.go

```diff
@@ -56,3 +58,252 @@ func TestCredentialAssertionResponse_json(t *testing.T) {
 		t.Errorf("Unmarshal() mismatch (-want +got):\n%s", diff)
 	}
 }
+
+func TestCredentialCreation_Validate(t *testing.T) {
+	okCC := &wanlib.CredentialCreation{
+		Response: protocol.PublicKeyCredentialCreationOptions{
+			Challenge: make([]byte, 32),
+			RelyingParty: protocol.RelyingPartyEntity{
+				ID: "example.com",
+			},
+			Parameters: []protocol.CredentialParameter{
+				{Type: protocol.PublicKeyCredentialType, Algorithm: webauthncose.AlgES256},
+			},
+			AuthenticatorSelection: protocol.AuthenticatorSelection{
+				UserVerification: protocol.VerificationDiscouraged,
+			},
+			Attestation: protocol.PreferNoAttestation,
+		},
+	}
+
+	pwdlessOK := *okCC
+	pwdlessOK.Response.RelyingParty.Name = "Teleport"
+	pwdlessOK.Response.User = protocol.UserEntity{
+		CredentialEntity: protocol.CredentialEntity{
+			Name: "llama",
+		},
+		DisplayName: "Llama",
+		ID:          []byte{1, 2, 3, 4, 5}, // arbitrary
+	}
+	rrk := true
+	pwdlessOK.Response.AuthenticatorSelection.RequireResidentKey = &rrk
+	pwdlessOK.Response.AuthenticatorSelection.UserVerification = protocol.VerificationRequired
+
+	tests := []struct {
+		name           string
+		createCC       func() *wanlib.CredentialCreation
+		alwaysCreateRK bool
+		wantErr        string
+	}{
+		{
+			name:     "ok", // check that good params are good
```

Comment:
This is all moved, right? Any changes?

Reply:
this one has some changes because now rrk checks are inlined